### ISSUE TITLE
fatal error: concurrent map read and map write

### ISSUE THREAD

**Issue by bkcsfi** [Jeffail/benthos#243]
Version: 2.11.0
Date: 2019-08-03T10:06:44Z

(also happens with 2.10)

[2015-01-benthos-crash-stack.log](https://github.com/Jeffail/benthos/files/3465466/2015-01-benthos-crash-stack.log)

[benthos_yaml.zip](https://github.com/Jeffail/benthos/files/3465483/benthos_yaml.zip)

I've attached what I'm sure is a pretty horrible yaml file that could be simplified. Its a process that reads some data from stdin, then uses subprocess a lot to retrieve data from a database (not postgres), find paths to documents and their types, convert the documents to normalized pdfs, concatenate them and write out a final pdf

What's interesting about this is that the crash occurs when logger level = INFO, but does not occur when logger level = DEBUG

I'm sure that's just random.

There's only one location where regex is used in the yaml file

I have not set any environment variables, so threads = 4 (also crashes with threads = 8, and threads = 1)

Also crashes when level=DEBUG and threads=1 .. 

this might be helpful. This graph is what I expect

```
Resolved DAG: [[get_ca_customs_document_path get_us_backup_document_path get_us_customs_document_path get_us_tif_backup_document_path get_ca_backup_document_path] [convert_us_tif_backup_document_path_to_pdf] [write_final_output_pdf]]
```

**Comment by bkcsfi** [comment]
I removed the regex from the config file, but the program still crashes with the same stack.

**Comment by Jeffail** [comment]
Hey @bkcsfi,

Taking a look. Would it be possible for you to also share the config snippets from: `./benthos_append_errors.yml#/pipeline/processors/0`?

One thing that immediately stands out is that multiple parallel stages in your DAG are setting `errors` and `warnings` in their postmap stages. This is a problem as Benthos sets those values by overriding the previous value, whereas what you'd want in this case is to append to the previous value.

One potential solution to this would be to refactor so that each stage, if an error occurs, populates a namespaced field within `errors`, which would be an object. So for example, it might look like this:

```json
{
  "errors": {
    "get_us_customs_document_path": [
      "foo", "bar", "baz"
    ]
  }
}
```

And your postmaps would look like this:

```yaml
postmap_optional:
  errors.get_us_customs_document_path: errors
```

Then you can check for total error counts with this jmespath:

```yaml
jmespath:
  query: 'length(values(errors)[]) > `0`'
```

**Comment by Jeffail** [comment]
Think I got it. It looks as though this was caused by the `process_map` processors that are nested within your DAG stages. They weren't correctly isolating their postmap stages so it opened up the possibility of race conditions versus parallel DAG stages. I've fixed the race condition in v2.11.1 so you shouldn't see the crash anymore. Can you give it a try?

Fixed: https://github.com/Jeffail/benthos/commit/b810be514147ac9d9b69ad129930f94c6d1531ce
Released: https://github.com/Jeffail/benthos/releases/tag/v2.11.1

**Comment by bkcsfi** [comment]
Hi,

Thanks for the quick fix. Version 2.11.1 does work now w/o error. I tried variations of thread count and log level and it worked every time.

Also thanks for the suggestion on how to handle merging lists from dag stages into a single 'list'. I hit that issue with temporary_documents, your solution is much nicer than mine.
